File: python-alert-service/messaging/consumer.py
"""RabbitMQ consumer for sensor update events using aio-pika (asyncio).

Reactive mode uses RxPY to model the event pipeline as an Observable stream
with backpressure via flat_map's max_concurrent parameter.
"""

import asyncio
import json
import logging

import aio_pika
import reactivex as rx
from reactivex import operators as ops
from reactivex.disposable import CompositeDisposable
from reactivex.scheduler.eventloop import AsyncIOScheduler
from reactivex.subject import Subject

from metrics import server as metrics_server

logger = logging.getLogger("alert_service")


class AlertConsumer:
    """Consumes sensor.updated events from RabbitMQ via aio-pika.

    Blocking mode (worker_count=0): each message is evaluated inline in the
    consumer coroutine before the next message is fetched — sequential,
    strong at-least-once.

    Reactive mode (worker_count>0): messages are emitted into an RxPY Subject.
    A flat_map operator fans out evaluation to up to ``worker_count`` concurrent
    asyncio tasks.  Backpressure: flat_map limits in-flight evaluations to
    ``worker_count``; when all slots are occupied, the Subject's on_next call
    awaits until a slot frees up.
    """

    def __init__(self, rabbitmq_url: str, callback, worker_count: int = 0):
        """Initialize the consumer.

        Args:
            rabbitmq_url: AMQP connection URL.
            callback: Async callable to invoke with each sensor event dict.
            worker_count: Number of concurrent reactive evaluations.
                          0 means blocking mode (no reactive pipeline).
        """
        self.rabbitmq_url = rabbitmq_url
        self.callback = callback
        self.worker_count = worker_count
        self._disposable = CompositeDisposable()

        if worker_count > 0:
            self._subject: Subject | None = Subject()
            self._mode = "reactive"
        else:
            self._subject = None
            self._mode = "blocking"

    def start(self):
        """Schedule the consumer loop as an asyncio task."""
        loop = asyncio.get_event_loop()

        if self._subject is not None:
            scheduler = AsyncIOScheduler(loop=loop)
            self._setup_reactive_pipeline(scheduler)

        loop.create_task(self._consume_loop())
        logger.info("Alert consumer started", extra={"mode": self._mode})

    def _setup_reactive_pipeline(self, scheduler):
        """Wire the RxPY Subject → flat_map → evaluate pipeline."""

        def evaluate_as_observable(event):
            """Wrap the async callback as a single-element Observable."""
            async def _run():
                try:
                    await self.callback(event)
                except Exception:
                    logger.exception("Reactive worker failed to evaluate event: %s", event)
                return event

            future = asyncio.ensure_future(_run())
            return rx.from_future(future)

        # map each event to an inner Observable, then merge with
        # max_concurrent to limit in-flight evaluations — this is the
        # reactive backpressure mechanism.
        pipeline = self._subject.pipe(
            ops.map(evaluate_as_observable),
            ops.merge(max_concurrent=self.worker_count),
        )

        disposable = pipeline.subscribe(
            on_next=lambda _: None,
            on_error=lambda e: logger.error("Reactive pipeline error: %s", e),
            scheduler=scheduler,
        )
        self._disposable.add(disposable)

    async def _consume_loop(self):
        """Reconnect loop — retries on connection failure."""
        while True:
            try:
                await self._consume()
            except Exception as e:
                logger.error(
                    "Consumer connection lost: %s — reconnecting in 5s", str(e)
                )
                await asyncio.sleep(5)

    async def _consume(self):
        """Connect to RabbitMQ and consume sensor.updated events."""
        connection = await aio_pika.connect_robust(self.rabbitmq_url)
        async with connection:
            channel = await connection.channel()

            # Prefetch limits how many unacked messages RabbitMQ sends at once.
            await channel.set_qos(prefetch_count=10)

            exchange = await channel.declare_exchange(
                "sensor_events", aio_pika.ExchangeType.FANOUT, durable=True
            )
            queue = await channel.declare_queue(
                "alert_service_python",
                durable=True,
                arguments={
                    "x-max-length": 1000,
                    "x-overflow": "reject-publish",
                },
            )
            await queue.bind(exchange)

            logger.info("Connected to RabbitMQ, waiting for sensor events")

            async with queue.iterator() as queue_iter:
                async for message in queue_iter:
                    await self._on_message(message)

    async def _on_message(self, message: aio_pika.IncomingMessage):
        """Handle incoming message."""
        try:
            event = json.loads(message.body)
        except json.JSONDecodeError:
            logger.warning("Received invalid JSON message")
            await message.nack(requeue=False)
            return

        if event.get("event") == "sensor.updated":
            if metrics_server.collector is not None:
                metrics_server.collector.inc_received()
            logger.info(
                "Received sensor.updated event",
                extra={
                    "sensor_id": event.get("sensor_id"),
                    "value": event.get("value"),
                    "trace_id": event.get("trace_id"),
                },
            )
            if self._subject is not None:
                # Reactive: emit into the RxPY stream; flat_map limits concurrency.
                self._subject.on_next(event)
            else:
                # Blocking: evaluate inline before acking.
                await self.callback(event)

        await message.ack()

File: go-alert-service/messaging/consumer.go
package messaging

import (
	"context"
	"encoding/json"
	"log/slog"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/reactivex/rxgo/v2"

	"iot-alert-service/metrics"
)

// SensorEvent represents a sensor.updated event from RabbitMQ.
type SensorEvent struct {
	Event     string  `json:"event"`
	SensorID  string  `json:"sensor_id"`
	Value     float64 `json:"value"`
	Type      string  `json:"type"`
	Unit      string  `json:"unit"`
	Timestamp string  `json:"timestamp"`
	TraceID   string  `json:"trace_id"`
}

// AlertConsumer consumes sensor.updated events from RabbitMQ.
type AlertConsumer struct {
	url      string
	callback func(SensorEvent)
	// Reactive pipeline fields; nil when running in blocking mode.
	eventCh     chan SensorEvent
	workerCount int
}

// NewAlertConsumer creates a new alert consumer in blocking mode.
func NewAlertConsumer(url string, callback func(SensorEvent)) *AlertConsumer {
	return &AlertConsumer{url: url, callback: callback}
}

// NewAsyncAlertConsumer creates a consumer that dispatches events through an
// RxGo reactive pipeline. Events are sent to a buffered channel; RxGo's
// FlatMap with WithPool(workerCount) fans out evaluation to concurrent
// goroutines. Backpressure: when the channel buffer is full, the producer
// blocks until a slot frees up.
func NewAsyncAlertConsumer(url string, callback func(SensorEvent), workerCount int) *AlertConsumer {
	ch := make(chan SensorEvent, workerCount*10)
	c := &AlertConsumer{url: url, callback: callback, eventCh: ch, workerCount: workerCount}
	c.startReactivePipeline()
	return c
}

// startReactivePipeline wires the channel into an RxGo Observable with
// FlatMap for concurrent evaluation.
func (c *AlertConsumer) startReactivePipeline() {
	// Adapt typed chan to chan rxgo.Item for FromChannel.
	itemCh := make(chan rxgo.Item, cap(c.eventCh))
	go func() {
		for event := range c.eventCh {
			itemCh <- rxgo.Of(event)
		}
		close(itemCh)
	}()

	// FlatMap fans out each event to a pool of goroutines.
	// WithPool limits concurrency — this is the reactive backpressure.
	observable := rxgo.FromChannel(itemCh).FlatMap(
		func(item rxgo.Item) rxgo.Observable {
			event := item.V.(SensorEvent)
			return rxgo.Defer([]rxgo.Producer{func(_ context.Context, next chan<- rxgo.Item) {
				func() {
					defer func() {
						if r := recover(); r != nil {
							slog.Error("Reactive worker panic recovered", "error", r, "sensor_id", event.SensorID)
						}
					}()
					c.callback(event)
				}()
				next <- rxgo.Of(event)
			}})
		}, rxgo.WithPool(c.workerCount),
	)

	// Subscribe in a background goroutine to drain the observable.
	go func() {
		<-observable.ForEach(
			func(_ interface{}) {},
			func(err error) { slog.Error("Reactive pipeline error", "error", err.Error()) },
			func() { slog.Info("Reactive pipeline completed") },
		)
	}()
}

// Start begins consuming in a background goroutine with reconnect logic.
func (c *AlertConsumer) Start() {
	mode := "blocking"
	if c.eventCh != nil {
		mode = "reactive"
	}
	go c.consumeLoop()
	slog.Info("Alert consumer started", "mode", mode)
}

func (c *AlertConsumer) consumeLoop() {
	for {
		err := c.consume()
		if err != nil {
			slog.Error("Consumer connection lost — reconnecting in 5s", "error", err.Error())
			time.Sleep(5 * time.Second)
		}
	}
}

func (c *AlertConsumer) consume() error {
	conn, err := amqp.Dial(c.url)
	if err != nil {
		return err
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()

	err = ch.ExchangeDeclare("sensor_events", "fanout", true, false, false, false, nil)
	if err != nil {
		return err
	}

	q, err := ch.QueueDeclare("alert_service_go", true, false, false, false, amqp.Table{
		"x-max-length": int32(1000),
		"x-overflow":   "reject-publish",
	})
	if err != nil {
		return err
	}

	err = ch.QueueBind(q.Name, "", "sensor_events", false, nil)
	if err != nil {
		return err
	}

	// Prefetch limits how many unacked messages RabbitMQ sends at once.
	err = ch.Qos(10, 0, false)
	if err != nil {
		return err
	}

	// autoAck=false: ack only after successful processing so messages are
	// not lost if the evaluator crashes mid-flight.
	msgs, err := ch.Consume(q.Name, "", false, false, false, false, nil)
	if err != nil {
		return err
	}

	slog.Info("Connected to RabbitMQ, waiting for sensor events")

	for msg := range msgs {
		var event SensorEvent
		if err := json.Unmarshal(msg.Body, &event); err != nil {
			slog.Warn("Received invalid JSON message", "error", err.Error())
			// Nack without requeue — malformed messages cannot be retried usefully.
			msg.Nack(false, false)
			continue
		}
		if event.Event == "sensor.updated" {
			metrics.EventsReceived.Add(1)
			slog.Info("Received sensor.updated event",
				"sensor_id", event.SensorID, "value", event.Value, "trace_id", event.TraceID)
			if c.eventCh != nil {
				// Reactive: send to RxGo pipeline; blocks if channel is full (backpressure).
				c.eventCh <- event
			} else {
				// Blocking: evaluate inline before acking.
				c.callback(event)
			}
		}
		msg.Ack(false)
	}

	return nil
}
